gr-qtgui/lib/fosphor_formatter_impl.cc

```diff
@@ -12,205 +13,223 @@
-        for (int i; i < items_to_process; i++) {
-            //// Update max hold:
-            // Compare with current max
-            volk_32f_x2_max_32f(
-                d_maxhold_buf.data(), d_maxhold_buf.data(), in_logfft_f, d_fft_size);
-            //// Update hit counter
+
+        // Update max
+        volk_32f_x2_max_32f(
+            d_maxhold_buf.data(), d_maxhold_buf.data(), d_logfft.data(), d_fft_size);
+
+        ++d_histo_count;
+        if (d_histo_count == d_histo_decim) {
+
+            ++n_histo;
+            d_histo_count = 0;
+
+            volk_32f_s32f_x2_convert_8u(
+                d_logfft_b.data(), d_logfft.data(), convert_mul, convert_add, d_fft_size);
+
+            // Update hit counter
             for (size_t i = 0; i < static_cast<size_t>(d_fft_size); i++) {
-                // This >>2 assumes d_num_bins is 64
-                const uint8_t bin_index = in_logfft_b[i] >> 2;
+                // TODO: allow num_bins other than 64 (implied here)
+                const uint8_t bin_index = d_logfft_b[i] >> 2;
                 d_hit_count[bin_index * d_fft_size + i]++;
             }
-            d_histo_count++;
-            in_logfft_f += d_fft_size;
-            in_logfft_b += d_fft_size;
-        }
-        consume_each(items_to_process);
-        if (d_histo_count < d_histo_decim) {
-            return 0;
-        }
-        d_histo_count = 0;
 
-        // Now update the output histogram buffer
-        for (size_t i = 0; i < d_histo_buf_f.size(); ++i) {
-            _update_histo_val(d_histo_buf_f[i], d_hit_count[i]);
-        }
+            // Now update the output histogram buffer
+            for (size_t i = 0; i < d_histo_buf_f.size(); ++i) {
 
-        // Copy the histogram buffer
-        float_array_to_uchar(d_histo_buf_f.data(), out, d_fft_size * d_num_bins, 256);
-        // Copy max hold to out buffer
-        const int maxhold_idx = d_num_bins * d_fft_size;
-        const int avg_idx = maxhold_idx + d_fft_size;
-        float_array_to_uchar(d_maxhold_buf.data(), out + maxhold_idx, d_fft_size);
-        // Copy average to out buffer
-        float_array_to_uchar(in_logfft_avg, out + avg_idx, d_fft_size);
-        const int num_items_written = d_num_bins + 2;
-        // Insert stream tag
-        auto tag = gr::tag_t{};
-        tag.key = pmt::string_to_symbol("rx_eob");
-        tag.value = pmt::PMT_T;
-        tag.offset = nitems_written(0) + num_items_written - 1;
-        add_item_tag(0, tag);
-
-        // Reset hit counter
-        std::fill(d_hit_count.begin(), d_hit_count.end(), 0);
-        return num_items_written;
-    }
+                const int16_t hc = d_hit_count[i];
+                float hv = d_histo_buf_f[i];
 
+                // Subtract from previous histogram values proportional to d_tdecay,
+                // and add in current hit count proportional to d_trise. Take
+                // decimation into account. (This code could be just a bit clearer).
```

Comment:
To be honest, I basically copy&pasted this from gr-fosphor, which is probably optimized for OpenCL.

Reply:
BTW, I think this function (updating a pixels luminescence, or whatever the right word is) could be pulled into its separate function, and then we could have fosphoric eye diagrams, constellation plots, etc.